WalletWasabi/Services/SingleInstanceChecker.cs

```diff
@@ -1,95 +1,156 @@
-		/// <param name="lockName">Name of system-wide mutex.</param>
-		public SingleInstanceChecker(Network network, string lockName)
+		public SingleInstanceChecker(int port)
 		{
-			Network = network;
-			_lockName = $"{MutexString}-{lockName}";
+			Port = port;
 		}
 
-		private IDisposable? SingleApplicationLockHolder { get; set; }
-		private Network Network { get; }
+		private int Port { get; }
 
-		public async Task CheckAsync()
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private TaskCompletionSource? TaskStartTcpListener { get; set; }
+
+		public event EventHandler? OtherInstanceStarted;
+
+		/// <summary>
+		/// This function ensures this is the first instance running on this machine or throws an exception if we are not. In case of secondary start
+		/// we are trying signal the first instance before throwing the exception.
+		/// On macOS this funtion will never throw if you run Wasabi as a macApp, because mac prevents running the same app multiple time on OS level.
+		/// </summary>
+		/// <exception cref="InvalidOperationException">Wasabi is already running, signalling the first instance failed.</exception>
+		/// <exception cref="OperationCanceledException">Wasabi is already running and signalled.</exception>
+		public async Task EnsureSingleOrThrowAsync()
 		{
-			if (_disposedValue)
+			if (DisposeCts.IsCancellationRequested)
 			{
 				throw new ObjectDisposedException(nameof(SingleInstanceChecker));
 			}
 
-			// The disposal of this mutex handled by AsyncMutex.WaitForAllMutexToCloseAsync().
-			var mutex = new AsyncMutex(_lockName);
 			try
 			{
-				using CancellationTokenSource cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(200));
-				SingleApplicationLockHolder = await mutex.LockAsync(cancellationToken: cts.Token).ConfigureAwait(false);
+				TaskStartTcpListener = new TaskCompletionSource();
+
+				// Start ExecuteAsync.
+				await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+
+				// Wait for the result of TcpListener.Start().
+				await TaskStartTcpListener.Task.WithAwaitCancellationAsync(DisposeCts.Token).ConfigureAwait(false);
+
+				// This is the first instance, nothing else to do.
+				return;
+			}
+			catch (SocketException ex) when (ex.ErrorCode is 10048 or 48 or 98)
+			{
+				// ErrorCodes are different on every OS: win, macOS, linux.
+				// It is already used -> another Wasabi is running on this network.
+				Logger.LogDebug("Detected another Wasabi instance.");
+			}
+
+			try
+			{
+				// Signal to the other instance, that there was an attempt to start the software.
+				using TcpClient client = new TcpClient();
+				using CancellationTokenSource timeoutCts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+				using CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(DisposeCts.Token, timeoutCts.Token);
+
+				await client.ConnectAsync(IPAddress.Loopback, Port, cts.Token).ConfigureAwait(false);
```

Comment:
If one's Localhost is different it does not matter for us because if you rin Wasabi on the same machine the client will still use the same. Does that make sense?

Reply:
Yes, i think you are right. IPv4 will never be turned off for local.